Allow record's ToString method to be sealed

thomaslevesque opened the pull request:
And don't generate ToString if base implementation is sealed

Fixes https://github.com/dotnet/csharplang/issues/4174

thomaslevesque commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -3644,16 +3644,30 @@ void addToStringMethod(MethodSymbol printMethod)
                 }
-                else
+
+                MethodSymbol? getBaseToStringMethod()
```
There's probably an easier way to find the base `ToString()` implementation, but I didn't find it... I'm open to suggestion.

jcouv replied:
`Compilation.GetRuntimeMember` and `WellKnownMembers.GetDescriptor` can help.

thomaslevesque replied:
I'll try using that, thanks

thomaslevesque replied:
Maybe I'm missing something, but the `WellKnownMember` enum doesn't seem to know about `Object.ToString`... How do I get a `MemberDescriptor` for it?

thomaslevesque replied:
@jcouv any idea how I could do this? Or do I just leave it as is?

jcouv replied:
Couldn't find anything better. Seems like that's what we have to do.

thomaslevesque commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -5895,14 +5895,7 @@ record C1
-            comp.VerifyEmitDiagnostics(
-                // (4,35): error CS8870: 'C1.ToString()' cannot be sealed because containing record is not sealed.
-                //     public sealed override string ToString() => throw null;
-                Diagnostic(ErrorCode.ERR_SealedAPIInRecord, "ToString").WithArguments("C1.ToString()").WithLocation(4, 35),
```
This is no longer illegal, which is the goal of this PR.

thomaslevesque commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -5895,14 +5895,7 @@ record C1
-                Diagnostic(ErrorCode.ERR_SealedAPIInRecord, "ToString").WithArguments("C1.ToString()").WithLocation(4, 35),
-                // (6,8): error CS0239: 'C2.ToString()': cannot override inherited member 'C1.ToString()' because it is sealed
-                // record C2 : C1;
-                Diagnostic(ErrorCode.ERR_CantOverrideSealed, "C2").WithArguments("C2.ToString()", "C1.ToString()").WithLocation(6, 8)
```
We no longer try to generate `ToString()` if the base implementation is sealed, so this diagnostic disappears.

thomaslevesque commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -6974,11 +6967,7 @@ extends System.Object
-            comp.VerifyEmitDiagnostics(
-                // (2,15): error CS0239: 'B.ToString()': cannot override inherited member 'A.ToString()' because it is sealed
-                // public record B : A {
-                Diagnostic(ErrorCode.ERR_CantOverrideSealed, "B").WithArguments("B.ToString()", "A.ToString()").WithLocation(2, 15)
```
We no longer try to generate ToString() if the base implementation is sealed, so this diagnostic disappears.

thomaslevesque commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -7013,11 +7002,7 @@ record C1
-            comp.VerifyEmitDiagnostics(
-                // (4,35): error CS8870: 'C1.ToString()' cannot be sealed because containing record is not sealed.
-                //     public sealed override string ToString() => throw null;
-                Diagnostic(ErrorCode.ERR_SealedAPIInRecord, "ToString").WithArguments("C1.ToString()").WithLocation(4, 35)
```
This is no longer illegal, which is the goal of this PR.

jcouv replied:
The new feature should be gated on a language version, so a diagnostic should remain for C# 9 in this scenario. The diagnostic will change though (it should say "to use this feature, you should use language version "preview"").

This means adding a new feature flag. Here's an existing example: `MessageID.IDS_FeatureLambdaDiscardParameters.CheckFeatureAvailability(...)`

thomaslevesque replied:
Ah, yes. Sorry, forgot about that (it's my first contribution on the compiler). I'll look into it.

thomaslevesque replied:
Done

thomaslevesque commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -15350,12 +15335,6 @@ public void Overrides_01()
                 Diagnostic(ErrorCode.ERR_MemberAlreadyExists, "Equals").WithArguments("Equals", "A").WithLocation(3, 33),
-                // (5,35): error CS8870: 'A.ToString()' cannot be sealed because containing record is not sealed.
-                //     public sealed override string ToString() => null;
-                Diagnostic(ErrorCode.ERR_SealedAPIInRecord, "ToString").WithArguments("A.ToString()").WithLocation(5, 35),
```
This is no longer illegal, which is the goal of this PR.

thomaslevesque commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -15350,12 +15335,6 @@ public void Overrides_01()
-                Diagnostic(ErrorCode.ERR_SealedAPIInRecord, "ToString").WithArguments("A.ToString()").WithLocation(5, 35),
-                // (7,8): error CS0239: 'B.ToString()': cannot override inherited member 'A.ToString()' because it is sealed
-                // record B(int X, int Y) : A
-                Diagnostic(ErrorCode.ERR_CantOverrideSealed, "B").WithArguments("B.ToString()", "A.ToString()").WithLocation(7, 8),
```
We no longer try to generate ToString() if the base implementation is sealed, so this diagnostic disappears.

thomaslevesque commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -15378,7 +15357,6 @@ public void Overrides_01()
                 "System.Int32 B.Y.get",
                 "void modreq(System.Runtime.CompilerServices.IsExternalInit) B.Y.init",
                 "System.Int32 B.Y { get; init; }",
-                "System.String B.ToString()",
```
`ToString()` is no longer generated when the base implementation is sealed.

jcouv replied:
This member should indeed be absent, both with language version 9 and preview.

thomaslevesque replied:
I don't get it. Why should it be absent in C# 9? Are you saying the existing test was wrong?

jcouv replied:
Yes, the existing test was wrong given the change to the language specification.

We should always produce the same symbols and bind the same way, regardless of language version. The only thing that can differ is what diagnostics are produced.
For example if you used `record struct` in an older language version, we'll still probably produce a symbol with members. But we'll also produce a diagnostic. When you switch to a newer language version, the error merely goes away.

Regarding the present test: the `ToString` member should never be produced when the base type has a sealed `ToString`. But an error should be produced when that happens with a language version which isn't high enough. This test should remain an error in C# 9, but the reason changes (it's no longer because we *can't* handle the situation, it's because we *won't* given the insufficient language version).

thomaslevesque replied:
> We should always produce the same symbols and bind the same way, regardless of language version. The only thing that can differ is what diagnostics are produced.

Ah, thanks, that answers a question I was going to ask: how to detect if a feature is enabled without emitting a diagnostic if it's not. And the answer is "I don't need to"...

thomaslevesque replied:
Done, I think

jcouv commented:
Please add an entry to `roslyn/docs/Language Feature Status.md` to document the existence and progress of the feature:
Feature = "sealed record ToString" + link to proposal
Branch = main (I think we won't need a feature branch for this, since single PR)
State = in progress + link to test plan (https://github.com/dotnet/roslyn/issues/52031)
Dev = you
Reviewer = jcouv
LDM champ = 333fred

Thanks

jcouv commented:
Thanks for taking this feature. Ping when ready for another look or if you have any questions.

thomaslevesque commented:
Some tests are failing, but they don't seem related to my changes... flaky tests? Or did I break something?

RikkiGibson commented:
Failures look flaky. Let's rerun the failing test jobs.

jcouv commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -5896,15 +5896,27 @@ record C1
         }
 
+        [Fact]
+        public void ToString_DerivedRecord_BaseHasSealedToString_PreviewSealedRecordToString()
```
nit: could be merged into preview test, since same source string (create a compilation and check diagnostics, then repeat with different options).
Please verify execution (using `CompileAndVerify(comp, expectedOutput: "...")`) as well to show that we didn't generate a new ToString.

thomaslevesque replied:
> could be merged into preview test

Do you mean changing the test to a `Theory` instead of `Fact`?

thomaslevesque replied:
> Also with some other ToString methods in base type (parameterlist not empty, return type not string)

I'm unable to introduce a `ToString()` method with a different return type. I get this error:
> error CS8869: 'C2.ToString()' does not override expected method from 'object'.

So I'll just add a method with non-empty parameter list.

(Not sure what this CS8869 error is, btw; I don't get this error if I do the same thing in a class instead of a record. What's the rationale behind this?)

jcouv replied:
> Do you mean changing the test to a Theory instead of Fact?

I had a typo. Meant into "previous" test:
```
string src = "...";
var comp = CreateCompilation(..., Regular9);
comp.VerifyEmitDiagnostics(...);
comp = CreateCompilation(..., RegularPreview);
comp.VerifyEmitDiagnostics();
CompileAndVerify(comp, expectedOutput: "..."); // can verify execution when no error in LangVer=preview
```

jcouv replied:
> I'm unable to introduce a ToString() method with a different return type.

It's possible, but requires jumping through some hoops by defining your own `System.Object` type. See `RecordTests.ObjectEquals_05` for help doing that. The key is to use `CreateEmptyCompilation` to avoid pulling libraries that already have this type.

thomaslevesque replied:
> I had a typo. Meant into "previous" test

Yes, I gathered as much.
Is it OK that I used a theory instead of creating compilations with different options in the same test?

thomaslevesque replied:
> It's possible, but requires jumping through some hoops by defining your own System.Object type. See RecordTests.ObjectEquals_05 for help doing that. The key is to use CreateEmptyCompilation to avoid pulling libraries that already have this type.

Thanks, I'll look into it. Not today, though...

thomaslevesque replied:
Added a test for this scenario (ToString_DerivedRecord_BaseBaseHasSealedToString_And_BaseHasToStringWithDifferentReturnType). I no longer needed to jump through hoops, thanks to the changes related to CS8869

jcouv commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -15350,12 +15446,63 @@ public void Overrides_01()
+                Diagnostic(ErrorCode.ERR_CantOverrideSealed, "B").WithArguments("B.GetHashCode()", "A.GetHashCode()").WithLocation(7, 8)
+                );
+
+            var actualMembers = comp.GetMember<NamedTypeSymbol>("B").GetMembers().ToTestDisplayStrings();
```
It'll be good to factor the C# 9 and preview tests into one so that  we can share the validation logic (showing that the expectations are the same).

jcouv reviewed (commented):
Done with review pass (iteration 4). It's looking good modulo some test compaction and testing a couple more scenarios.

jcouv commented:
Just a heads up on process for compiler features:
- as any compiler PR, we need two reviews
- although I don't expect much risk for this feature, we'll need IDE sign-off as well. I'll find an IDE buddy to represent
- the team will a feature review to brainstorm some test (I'll start email later this week). Jared had a [suggestion](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQGYACLEgYRIG8CT6zSBnAUwgBsWESB7ANxYAnQVAQsy2AAwkAKjwDKAFxEwA5gAoAlCQC8APhIAiYOwgALQwG4CAXwIFyAERIhKNOg2IkYLAO4k+KEFFAFcOCWk5JRUNbX0jE3MrWyA=) already.

thomaslevesque commented:
@jcouv I think I addressed all comments from your last review
I merged all "_PreviewSealedRecordToString" tests with the C#9 tests using Theory and InlineData.
I added the required tests (to cover sealed `ToString` in base's base, and `ToString` with non-empty parameter list in base)

> * the team will a feature review to brainstorm some test (I'll start email later this week). Jared had a [suggestion](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQGYACLEgYRIG8CT6zSBnAUwgBsWESB7ANxYAnQVAQsy2AAwkAKjwDKAFxEwA5gAoAlCQC8APhIAiYOwgALQwG4CAXwIFyAERIhKNOg2IkYLAO4k+KEFFAFcOCWk5JRUNbX0jE3MrWyA=) already.

Ah, yes, that could be a problem... but if you try to [do the same with records instead of classes](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQGYACAJwFMBjAezIRIGESBvAkjk4kmgNwrJkoCCl2wAGEgBUaAZQAuQmAHMAFAEoSAXgB8JAETAANhAAW+gNwEAvgQKVa9EgBESIJq3aduMCgHcSXigyeQBXCCMxSRkFJTVNXQNjM0sbIA==), you get this:

> error CS8869: 'D.ToString()' does not override expected method from 'object'.

So I don't think this situation can actually occur

jcouv commented:
> Ah, yes, that could be a problem... but if you try to do the same with records instead of classes, you get this

It should become possible in records with this PR.

jcouv reviewed (approved):
LGTM Thanks (iteration 5)

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -6600,4 +6600,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
     <value>Using a function pointer type in a 'typeof' in an attribute is not supported.</value>
   </data>
+  <data name="IDS_FeatureSealedRecordToString" xml:space="preserve">
+    <value>sealed record ToString</value>
```
>sealed record ToString [](start = 11, length = 22)

"sealed ToString in record"?

thomaslevesque replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Portable/Errors/MessageID.cs:
```diff
@@ -324,6 +325,7 @@ internal static LanguageVersion RequiredVersion(this MessageID feature)
             {
                 // C# preview features.
                 case MessageID.IDS_FeatureMixedDeclarationsAndExpressionsInDeconstruction:
+                case MessageID.IDS_FeatureSealedRecordToString:
```
>case MessageID.IDS_FeatureSealedRecordToString: [](start = 16, length = 47)

I am assuming this is a semantic check. Consider adding a comment.

thomaslevesque replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -3644,16 +3644,41 @@ void addToStringMethod(MethodSymbol printMethod)
+                            .SingleOrDefault(m => m.ParameterCount == 0 && m.ReturnType.SpecialType == SpecialType.System_String);
+
+                        if (method is not null)
+                            return method;
```
>return method; [](start = 28, length = 14)

The fact that the name and the signature match doesn't mean this is an override of Object.ToString. We should be able to come up with a test scenario with base like that. If not defined in source, then in IL.

AlekseyTs replied:
I think we should mimic the check that is performed by ```SynthesizedRecordObjectMethod.VerifyOverridesMethodFromObject```, but I would also simplify its implementation by switching to GetSpecialTypeMember helper to locate the Object member and comparing it with ```leastOverridden``` instead of performing manual signature comparison.

---
In reply to: [601500624](https://github.com/dotnet/roslyn/pull/52029#discussion_r601500624) [](ancestors = 601500624)

AlekseyTs replied:
FYI, I just merged refactoring for ```SynthesizedRecordObjectMethod.VerifyOverridesMethodFromObject``` to take advantage of ```SpecialMemeber```.

---
In reply to: [601544539](https://github.com/dotnet/roslyn/pull/52029#discussion_r601544539) [](ancestors = 601544539,601500624)

thomaslevesque replied:
@AlekseyTs I made some changes to `getBaseToStringMethod` (compared least overriden to object.ToString(), replaced Linq usage with a loop). Is this what you had in mind?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -3644,16 +3644,41 @@ void addToStringMethod(MethodSymbol printMethod)
                 else
                 {
                     var toStringMethod = (MethodSymbol)existingToStringMethod;
                     if (!SynthesizedRecordObjectMethod.VerifyOverridesMethodFromObject(toStringMethod, SpecialType.System_String, diagnostics) && toStringMethod.IsSealed && !IsSealed)
```
I mean in cases when the override of Object.ToString is sealed up the hierarchy.

---
In reply to: [601507591](https://github.com/dotnet/roslyn/pull/52029#discussion_r601507591) [](ancestors = 601507591)

thomaslevesque replied:
@AlekseyTs

> If we allow sealing Object.ToString up the hierarchy, it feels like we should allow to declare new ToString method with the same parameter types that doesn't override Object.ToString, but shadows it instead.

Why is it currently forbidden in records anyway?
Anyway, wouldn't it be out of scope for this PR? I set out to allow sealing `ToString()`, not remove a restriction I didn't even know about...

AlekseyTs replied:
>> If we allow sealing Object.ToString up the hierarchy, it feels like we should allow to declare new ToString method with the same parameter types that doesn't override Object.ToString, but shadows it instead.
>> I mean in cases when the override of Object.ToString is sealed up the hierarchy.

>Why is it currently forbidden in records anyway?
>Anyway, wouldn't it be out of scope for this PR? I set out to allow sealing ToString(), not remove a restriction I didn't even know about...

This wasn't specifically forbidden. However, there was a requrement that a ```string ToString()``` method added explicitly overrides ```Object.ToString```. We are lifting this requirement for cases when ```Object.ToString``` is sealed in a base class. Therefore, when one adds ```string ToString()``` method in a situation like that, compiler shouldn't require that that method overrides ```Object.ToString```, because this simply cannot be done. It is not posible to override a sealed method. Please create a unit-test for this scenaro, I think that should give you the idea what I am talking about. 


---
In reply to: [602653210](https://github.com/dotnet/roslyn/pull/52029#discussion_r602653210) [](ancestors = 602653210)

thomaslevesque replied:
> Please create a unit-test for this scenaro, I think that should give you the idea what I am talking about.

@AlekseyTs do you mean something like this?

```csharp
        [Fact]
        public void ToString_NewToString_SealedBaseToString()
        {
            var source = @"
public record A
{
    public sealed override string ToString() => throw null;
}

public record B : A
{
    public new string ToString() => throw null;
}";

            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
            comp.VerifyEmitDiagnostics();
        }
```

This currently fails with this error:

```
// (9,23): error CS8869: 'B.ToString()' does not override expected method from 'object'.
//     public new string ToString() => throw null;
```

Should I make this test pass, then?

AlekseyTs replied:
> do you mean something like this?

Yes.

> Should I make this test pass, then?

I think yes.

---
In reply to: [602934953](https://github.com/dotnet/roslyn/pull/52029#discussion_r602934953) [](ancestors = 602934953)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -3644,16 +3644,41 @@ void addToStringMethod(MethodSymbol printMethod)
+                    while (tmp is not null)
+                    {
+                        var method = tmp.GetSimpleNonTypeMembers(WellKnownMemberNames.ObjectToString)
+                            .OfType<MethodSymbol>()
```
>OfType [](start = 29, length = 6)

I think a regular loop through members would be preferred other the Linq query in this case.

thomaslevesque replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -6973,12 +7028,11 @@ extends System.Object
+            var comp = CreateCompilationWithIL(
+                new[] { source, IsExternalInitTypeDefinition },
+                ilSource: ilSource,
+                parseOptions: usePreview ? TestOptions.RegularPreview : TestOptions.Regular9);
```
>TestOptions.Regular9 [](start = 72, length = 20)

I think we should report a language version error for `Regular9` because previous compiler wouldn't be able to compile this code and targeting that version should prevent a build break like that when compiler is switched back.

thomaslevesque replied:
Sorry, maybe I'm a bit slow (it's 3 a.m. here), but I don't get it... Could you elaborate?

AlekseyTs replied:
>> I think we should report a language version error for Regular9 because previous compiler wouldn't be able to compile this code and targeting that version should prevent a build break like that when compiler is switched back.

> Sorry, maybe I'm a bit slow (it's 3 a.m. here), but I don't get it... Could you elaborate?

There is a requirement - If compiling some code with a particular language version succeeds, then compiling the same code by using an old version of the compiler that added support for that language version should also succeed and result should be equivalent. This is an error scenario in the previous version of the compiler. Therefore, when we compile this code with language version C# 9, we should report an error, saying that the language version should be bumped up. 

---
In reply to: [602653525](https://github.com/dotnet/roslyn/pull/52029#discussion_r602653525) [](ancestors = 602653525)

thomaslevesque replied:
So, just to make sure I understand it right... If language version is 9, it should:
- not emit a `ToString()` override (since the method is sealed in the base type)
- emit a `ERR_FeatureInPreview` diagnostic

Is that correct?

I just worry that the error will be a bit confusing. The user would just be trying to inherit a record with a sealed ToString (e.g. from another assembly compiled with C# 10), but the error will say that "sealed ToString is records" is a preview feature.

AlekseyTs replied:
>So, just to make sure I understand it right... If language version is 9, it should:
>- not emit a ToString() override (since the method is sealed in the base type)
>- emit a ERR_FeatureInPreview diagnostic
>
>Is that correct?
>
>I just worry that the error will be a bit confusing. The user would just be trying to inherit a record with a sealed ToString (e.g. from another assembly compiled with C# 10), but the error will say that "sealed ToString is records" is a preview feature.

If the simple error will be confusing, we should consider adding a dedicated more descriptive error, which will request to bump the language version. For example: "Inheriting from a class with a sealed 'Object.ToString' is not supported in C# 9. Please use language version 'preview' or greater."

---
In reply to: [602872811](https://github.com/dotnet/roslyn/pull/52029#discussion_r602872811) [](ancestors = 602872811)

thomaslevesque replied:
OK, I addressed that in commit 7dcf8150e63. (also added the new error message)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -5883,26 +5883,79 @@ public void ToString_DerivedRecord_AbstractSealed()
+
         [Fact]
-        public void ToString_DerivedRecord_BaseHasSealedToString()
+        public void ToString_DerivedRecord_BaseBaseHasSealedToString()
```
>ToString_DerivedRecord_BaseBaseHasSealedToString [](start = 20, length = 48)

I think we should add more tests for:
- trying to override the sealed method explicitly and how next level derived records are going to behave.
- trying to shadow method, including in a way when it will not be accessible in derived types 
- inheritance behvior in presence of shadowing up the hierarhy

thomaslevesque replied:
> * trying to override the sealed method explicitly and how next level derived records are going to behave.

Done in ToString_DerivedRecord_BaseBaseHasSealedToString_And_BaseTriesToOverride

> * trying to shadow method, including in a way when it will not be accessible in derived types, including when shadowed method is not sealed.

Done in ToString_DerivedRecord_BaseBaseHasSealedToString_And_BaseShadowsToStringPrivate and ToString_DerivedRecord_BaseBaseHasSealedToString_And_BaseShadowsToStringNonSealed

> * inheritance behvior in presence of shadowing up the hierarhy

I think this is already covered in ToString_NewToString_SealedBaseToString

AlekseyTs commented:
Done with review pass (commit 5)

thomaslevesque commented:
> You're right, it'll still be an error. Let's add the test please

@jcouv sorry, I don't get it. What test should I add?

thomaslevesque commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -3644,16 +3644,41 @@ void addToStringMethod(MethodSymbol printMethod)
+                    {
+                        var method = tmp.GetSimpleNonTypeMembers(WellKnownMemberNames.ObjectToString)
+                            .OfType<MethodSymbol>()
+                            .SingleOrDefault(m => m.ParameterCount == 0 && m.ReturnType.SpecialType == SpecialType.System_String);
```
@AlekseyTs the new matching logic (using least overridden) accounts for that, right?

AlekseyTs commented:
@thomaslevesque It looks like you rewrote commit in this PR. In the future, please don't rewrite commits in a PR under review. This confuses review tool that we are using and often prevents us from doing incremental reviews. Simply make changes and push new commits.

thomaslevesque commented:
> It looks like you rewrote commits in this PR. In the future, please don't rewrite commits in a PR under review. This confuses review tool that we are using and often prevents us from doing incremental reviews. Simply make changes and push new commits.

Ah, sorry, I didn't realize it would cause issues. I just rebased onto main to get your refactored VerifyOverridesMethodFromObject. Probably should have merged, but I'm used to rebasing. I'll keep that in mind in the future.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -3641,17 +3641,54 @@ void addToStringMethod(MethodSymbol printMethod)
+                MethodSymbol? getBaseToStringMethod()
+                {
+                    var objectToString = this.DeclaringCompilation.GetSpecialTypeMember(SpecialMember.System_Object__ToString);
+                    var tmp = this.BaseTypeNoUseSiteDiagnostics;
```
>tmp [](start = 24, length = 3)

Please use more descriptive name for the local

thomaslevesque replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -6600,4 +6600,10 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
+    <value>sealed ToString in record</value>
+  </data>
+  <data name="ERR_InheritingFromRecordWithSealedToString" xml:space="preserve">
+    <value>Inheriting from a record with a sealed 'Object.ToString' is not supported in C# 9. Please use language version 'preview' or greater.</value>
```
>Inheriting from a record with a sealed 'Object.ToString' is not supported in C# 9. Please use language version 'preview' or greater. [](start = 11, length = 132)

We don't hardcode language versions for errors like this. For example, see how we encorpoarte language version information into ERR_InvalidModifierForLanguageVersion.  
We also need to make sure that IDE will offer upgrading the language version for the progect based on the error, it looks like some changes should be made in src\Features\CSharp\Portable\UpgradeProject\CSharpUpgradeProjectCodeFixProvider.cs to make this work. Appropriate test(s) should be added in src\EditorFeatures\CSharpTest\Diagnostics\UpgradeProject\UpgradeProjectTests.cs to verify the fact.

thomaslevesque replied:
I changed the error message to use placeholders, but I haven't made the changes on project upgrade yet.

thomaslevesque replied:
> We also need to make sure that IDE will offer upgrading the language version for the progect based on the error, it looks like some changes should be made in src\Features\CSharp\Portable\UpgradeProject\CSharpUpgradeProjectCodeFixProvider.cs to make this work. Appropriate test(s) should be added in src\EditorFeatures\CSharpTest\Diagnostics\UpgradeProject\UpgradeProjectTests.cs to verify the fact.

This is proving more difficult than I anticipated... The implementation is trivial (just add CS8912 to the `FixableDiagnosticIds` property), but I'm unable to test it.

There are 2 errors related to this feature:
- CS8652 (ERR_FeatureInPreview), which is already covered by the tests, but isn't specific to this feature
- CS8912 (ERR_InheritingFromRecordWithSealedToString)

In order to trigger CS8912, the base record must be declared in a different module, but the current test infrastructure doesn't allow for that. I'd change it, but I'm not familiar with the workspace API and I don't really know where to start...

AlekseyTs replied:
> In order to trigger CS8912, the base record must be declared in a different module, but the current test infrastructure doesn't allow for that. I'd change it, but I'm not familiar with the workspace API and I don't really know where to start...

Actually it looks like test infatrsucture supports testing scenarios when more than one compilation is involved. See UpgradeProjectForDefaultInterfaceImplementation_CS8706 for example.

---
In reply to: [605998135](https://github.com/dotnet/roslyn/pull/52029#discussion_r605998135) [](ancestors = 605998135)

thomaslevesque replied:
Thanks, I'm looking into it

thomaslevesque replied:
OK, done

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -5883,26 +5883,82 @@ public void ToString_DerivedRecord_AbstractSealed()
+                    // (4,35): error CS8652: The feature 'sealed ToString in record' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                    //     public sealed override string ToString() => throw null;
+                    Diagnostic(ErrorCode.ERR_FeatureInPreview, "ToString").WithArguments("sealed ToString in record").WithLocation(7, 35),
+                    // (9,8): error CS8912: Inheriting from a record with a sealed 'Object.ToString' is not supported in C# 9. Please use language version 'preview' or greater.
```
>// (9,8): error CS8912: Inheriting from a record with a sealed 'Object.ToString' is not supported in C# 9. Please use language version 'preview' or greater. [](start = 20, length = 156)

In this scenario the error feels somewhat redundant. In cases like this, we usually supress it. For example, if the base that seals Object.ToStrung is declared in the same module, we can skip reporting the error. VerifyOverridesMethodFromObject helper does something similar.

thomaslevesque replied:
Done.
I'd like to add a test to ensure that the warning is emitted if the base that seals `ToString` *isn't* in the same module, but I'm not sure how to do that... I could probably do it with IL source, but that seems overkill; maybe there's an easier way?

AlekseyTs replied:
>I could probably do it with IL source, but that seems overkill; maybe there's an easier way?

It is possible to reference one compilation from another. See Inheritance_03 unit-test for example.

---
In reply to: [604865992](https://github.com/dotnet/roslyn/pull/52029#discussion_r604865992) [](ancestors = 604865992)

thomaslevesque replied:
This scenario is covered in `ToString_BadBase_SealedToString`

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -15345,24 +15424,41 @@ public void Overrides_01()
+                    // (7,8): error CS0239: 'B.GetHashCode()': cannot override inherited member 'A.GetHashCode()' because it is sealed
+                    // record B(int X, int Y) : A
+                    Diagnostic(ErrorCode.ERR_CantOverrideSealed, "B").WithArguments("B.GetHashCode()", "A.GetHashCode()").WithLocation(7, 8),
+                    // (7,8): error CS8912: Inheriting from a record with a sealed 'Object.ToString' is not supported in C# 9. Please use language version 'preview' or greater.
```
>// (7,8): error CS8912: Inheriting from a record with a sealed 'Object.ToString' is not supported in C# 9. Please use language version 'preview' or greater. [](start = 20, length = 156)

Similarly, the error feels redundant here.

thomaslevesque replied:
Done

AlekseyTs commented:
Done with review pass (commit 11)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -3639,18 +3639,67 @@ void addToStringMethod(MethodSymbol printMethod)
                 {
-                    var toStringMethod = new SynthesizedRecordToString(this, printMethod, memberOffset: members.Count, diagnostics);
-                    members.Add(toStringMethod);
+                    if (baseToStringMethod?.ContainingModule != this.ContainingModule)
```
>? [](start = 42, length = 1)

It feels like the conditional access is not necessary here.

thomaslevesque replied:
I know, but the compiler disagrees (CS8602 Dereference of a possible null reference). But you're right, I should use `!` instead of `?`, since I know it's not null.

thomaslevesque replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -3639,18 +3639,67 @@ void addToStringMethod(MethodSymbol printMethod)
+                var baseToStringMethod = getBaseToStringMethod();
+                bool isBaseToStringSealed = baseToStringMethod is { IsSealed: true };
+
+                if (!this.DeclaringCompilation.IsFeatureEnabled(MessageID.IDS_FeatureSealedToStringInRecord) && isBaseToStringSealed)
```
>isBaseToStringSealed [](start = 112, length = 20)

Consider checking this first. In fact, it looks like the code can be adjusted to check this only once. Something like:
```
if (isBaseToStringSealed)
{
                    if (baseToStringMethod?.ContainingModule != this.ContainingModule && !this.DeclaringCompilation.IsFeatureEnabled(MessageID.IDS_FeatureSealedToStringInRecord))
                    {
                        var languageVersion = ((CSharpParseOptions)this.Locations[0].SourceTree!.Options).LanguageVersion;
                        var requiredVersion = MessageID.IDS_FeatureSealedToStringInRecord.RequiredVersion();
                        diagnostics.Add(
                            ErrorCode.ERR_InheritingFromRecordWithSealedToString,
                            this.Locations[0],
                            languageVersion.ToDisplayString(),
                            new CSharpRequiredLanguageVersion(requiredVersion));
                    }
}
else
{
                    if (!memberSignatures.TryGetValue(targetMethod, out Symbol? existingToStringMethod))
                    {
                            var toStringMethod = new SynthesizedRecordToString(this, printMethod, memberOffset: members.Count, diagnostics);
                            members.Add(toStringMethod);
                    }
                    else
                    {
                        var toStringMethod = (MethodSymbol)existingToStringMethod;
                            if (!SynthesizedRecordObjectMethod.VerifyOverridesMethodFromObject(toStringMethod, SpecialMember.System_Object__ToString, diagnostics) && toStringMethod.IsSealed && !IsSealed)
                            {
                                MessageID.IDS_FeatureSealedToStringInRecord.CheckFeatureAvailability(
                                    diagnostics,
                                    this.DeclaringCompilation,
                                    toStringMethod.Locations[0]);
                            }
                    }
}
```

thomaslevesque replied:
Indeed, it's much cleaner that way. After a few rounds of changes the logic was becoming difficult to follow...

thomaslevesque replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -3639,18 +3639,67 @@ void addToStringMethod(MethodSymbol printMethod)
-                    if (!SynthesizedRecordObjectMethod.VerifyOverridesMethodFromObject(toStringMethod, SpecialMember.System_Object__ToString, diagnostics) && toStringMethod.IsSealed && !IsSealed)
+                    if (!memberSignatures.TryGetValue(targetMethod, out Symbol? existingToStringMethod))
+                    {
+                        if (!isBaseToStringSealed)
```
>if (!isBaseToStringSealed) [](start = 24, length = 26)

Could this be moved out of the `if` as suggested above?

thomaslevesque replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -3639,18 +3639,67 @@ void addToStringMethod(MethodSymbol printMethod)
+                    else
+                    {
+                        var toStringMethod = (MethodSymbol)existingToStringMethod;
+                        if (!isBaseToStringSealed)
```
>if (!isBaseToStringSealed) [](start = 24, length = 26)

Could this be moved out of the `if` as suggested above?

thomaslevesque replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -6904,7 +6957,27 @@ extends System.Object
+}";
+
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyEmitDiagnostics();
```
>comp.VerifyEmitDiagnostics(); [](start = 12, length = 29)

For success scenarios like this we prefer to compile, execute the compiled code and observe that it behaves as expected at runtime. For example in this case, it would be good to verify that calling B.ToString calls the method from B, but calling ((A)B).ToString calls the method from A. The methods can return distinct values and we can print them and verify the output.

thomaslevesque replied:
OK, I'll change it

thomaslevesque replied:
Done (also in ToString_BadBase_SealedToString)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -6973,12 +7046,22 @@ extends System.Object
+            else
+            {
+                comp.VerifyEmitDiagnostics(
+                    // (2,1): error CS8912: Inheriting from a record with a sealed 'Object.ToString' is not supported in C# 9. Please use language version 'preview' or greater.
```
>// (2,1): error CS8912: Inheriting from a record with a sealed 'Object.ToString' is not supported in C# 9. Please use language version 'preview' or greater. [](start = 20, length = 156)

Consider adjusting the message to the new form.

thomaslevesque replied:
Do you mean with "9.0" instead of "9"? Or using the placeholders?

thomaslevesque replied:
Done, if you just meant "9.0" insted of "9"

AlekseyTs commented:
Done with review pass (commit 15)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -3639,18 +3639,61 @@ void addToStringMethod(MethodSymbol printMethod)
+                var baseToStringMethod = getBaseToStringMethod();
+                bool isBaseToStringSealed = baseToStringMethod is { IsSealed: true };
+
+                if (isBaseToStringSealed)
```
>isBaseToStringSealed [](start = 20, length = 20)

Since this local is used only in one place, consider inlining ```baseToStringMethod is { IsSealed: true }``` condition here, then we wouldn't need to supress nullable warnings when ```baseToStringMethod``` is dereferenced.

thomaslevesque replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -6965,20 +7141,33 @@ extends System.Object
 ";
             var source = @"
+var b = new B();
+System.Console.Write(b.ToString());
```
Did you mean to execute code in success branch below?

thomaslevesque replied:
Oops, yes I did.

thomaslevesque replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -5883,26 +5883,176 @@ public void ToString_DerivedRecord_AbstractSealed()
+";
+
+            var comp = CreateCompilation(src, parseOptions: TestOptions.RegularPreview, options: TestOptions.DebugExe);
+            comp.VerifyEmitDiagnostics();
```
>comp.VerifyEmitDiagnostics(); [](start = 12, length = 29)

I think it will be good to assert that there is no ToString method in C3.

thomaslevesque replied:
Will do

thomaslevesque replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:
```diff
@@ -5883,26 +5883,176 @@ public void ToString_DerivedRecord_AbstractSealed()
+";
+
+            var comp = CreateCompilation(src, parseOptions: TestOptions.RegularPreview, options: TestOptions.DebugExe);
+            comp.VerifyEmitDiagnostics();
```
>comp.VerifyEmitDiagnostics(); [](start = 12, length = 29)

I think it will be good to assert that there is no ToString method in C3.

thomaslevesque replied:
Done

AlekseyTs commented:
Done with review pass (commit 19)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -3639,18 +3639,60 @@ void addToStringMethod(MethodSymbol printMethod)
                 {
-                    var toStringMethod = new SynthesizedRecordToString(this, printMethod, memberOffset: members.Count, diagnostics);
-                    members.Add(toStringMethod);
+                    if (baseToStringMethod!.ContainingModule != this.ContainingModule && !this.DeclaringCompilation.IsFeatureEnabled(MessageID.IDS_FeatureSealedToStringInRecord))
```
>! [](start = 42, length = 1)

Is the suppression still needed?

thomaslevesque replied:
No, it's not. I removed it.

AlekseyTs commented on src/EditorFeatures/CSharpTest/Diagnostics/UpgradeProject/UpgradeProjectTests.cs:
```diff
@@ -1003,5 +1004,31 @@ interface I2<out T1>
+        [Fact]
+        public async Task UpgradeProjectForSealedToStringInRecords_CS8912()
+        {
+            await TestLanguageVersionUpgradedAsync(
```
>TestLanguageVersionUpgradedAsync [](start = 18, length = 32)

Did you confirm that the test fails without chnages in CSharpUpgradeProjectCodeFixProvider?

thomaslevesque replied:
Yes. It failed with "No action was offered when one was expected"

AlekseyTs reviewed (approved):
LGTM (commit 23) with a couple of questions.

AlekseyTs commented:
@jcouv Please review the latest revision.

AlekseyTs commented:
@dotnet/roslyn-ide Please review IDE side of changes

AlekseyTs reviewed (approved):
LGTM (commit 4)

jcouv reviewed (approved):
LGTM Thanks (iteration 24)

jcouv commented:
/azp run

jcouv commented:
Merged/squashed. Thanks @thomaslevesque for the contribution!

thomaslevesque commented:
Thanks for the review @jcouv and @AlekseyTs!
